### DISCUSSION TITLE
Conditional Trust Email for OIDC, Google and GitLab Identity Provider

### DISCUSSION THREAD

**Discussion by cgeorgilakis** [keycloak/keycloak#8622] [Ideas]
Keycloak already supports skipping email verification for all users registering from a specific IdP by enabling the "Trust email" configuration option for that IdP. However, there is no support for conditionally skipping email verification for a given IdP based on the email information released during registration. Specifically, it should be possible to skip the verification process if the user's authenticating IdP releases an already verified email address. In the case of OpenID Connect Providers , Google and GitLab this can be determined based on the availability of the email claim combined with an email_verified claim set to true (defined in the standard OIDC claims). Maybe other Social Identity Providers have the same - or similar- functionality.
Could you review the [related PR](https://github.com/keycloak/keycloak/pull/8018) in order to add these functionality to Keycloak?
More information can be found to [related jira issue](https://issues.redhat.com/browse/KEYCLOAK-17851).

**Comment by stianst** [comment]
This is a perfectly valid use-case, so discussion is more around how to implement it, which we can talk about through the PR.

**Comment by cgeorgilakis** [comment]
I see that you have closed [jira issue](https://issues.redhat.com/browse/KEYCLOAK-17851) for this needed extension for emailVerified.

Based on your and @mposolda comment in the [PR](https://github.com/keycloak/keycloak/pull/8018), I have extended Attribute mappers ( OIDC, SAML , Social) for supporting emailVerified user field in this [branch](https://github.com/eosc-kc/keycloak/tree/70-verified-mapper).

I propose to write a more suitable github issue based on new implementation( emailVerified enhancement in SAML, OIDC, Social Attribe mapper) and create a new PR keeping relation with old jira issue and old PR.

@stianst @mposolda do you agree? Could you review the new PR whenever possible?

**Comment by larsw** [comment]
This should really be supported

**Comment by StarpTech** [comment]
Any updates on this? Would be great to hear an opinion from the project maintainers. I see this as a big shortcoming when dealing with many idp dynamically.

**Comment by extremegf** [comment]
I also want to enable the logic, where keycloak email check is skipped only if IdP says email_verified: true.
This is neccessary for Login with Microsoft, as not all emails from them are verified. 

I see that the discussion on adding "Conditional skip email verification" option, ended with no conclusion.

Does anyone know if there's currently a workaround for this? Maybe some custom SPI code?

**Reply by jb204** [reply]
Ended up putting together a custom mapper because the discord social IDP plugin was only going to let me import/hardcode attributes. From what I can see it shouldn't require too much to have the functionality moved into either the social IDP plugin itself or the generic OIDC.
As has been noted above the problem is that by having to do it this way, targeting a different claim as your source of 'email verified' truth means having to run maven, not just edit a admin panel text field and hit save.

**Reply by extremegf** [reply]
I might actually as an LLM to just write the whole project based on some other simple plugins from github.
